packages/mdc-ripple/_mixins.scss

```diff
@@ -21,68 +21,123 @@
 //
 
 @import "@material/animation/variables";
+@import "@material/base/mixins";
+@import "@material/feature-targeting/functions";
+@import "@material/feature-targeting/mixins";
 @import "@material/theme/mixins";
 @import "./functions";
+@import "./keyframes";
 @import "./variables";
 
-@mixin mdc-ripple-surface() {
-  --mdc-ripple-fg-size: 0;
-  --mdc-ripple-left: 0;
-  --mdc-ripple-top: 0;
-  --mdc-ripple-fg-scale: 1;
-  --mdc-ripple-fg-translate-end: 0;
-  --mdc-ripple-fg-translate-start: 0;
+@mixin mdc-ripple-common($query: mdc-feature-all()) {
+  $feat-animation: mdc-feature-create-target($query, animation);
+  $feat-structure: mdc-feature-create-target($query, structure);
 
-  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
-  will-change: transform, opacity;
+  // Ensure that styles needed by any component using MDC Ripple are emitted, but only once.
+  // (Every component using MDC Ripple imports these mixins, but doesn't necessarily import
+  // mdc-ripple.scss.)
+  @include mdc-feature-targets($feat-animation) {
+    @include mdc-base-emit-once("mdc-ripple/common/animation") {
+      @include mdc-ripple-keyframes_;
+    }
+  }
 
-  &::before,
-  &::after {
-    position: absolute;
-    border-radius: 50%;
-    opacity: 0;
-    pointer-events: none;
-    content: "";
+  @include mdc-feature-targets($feat-structure) {
+    @include mdc-base-emit-once("mdc-ripple/common/structure") {
+      // Styles used to detect buggy behavior of CSS custom properties in Edge.
+      // See: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/11495448/
+      // This is included in _mixins.scss rather than mdc-ripple.scss so that it will be
+      // present for other components which rely on ripple as well as mdc-ripple itself.
+      .mdc-ripple-surface--test-edge-var-bug {
+        --mdc-ripple-surface-test-edge-var: 1px solid #000;
+
+        visibility: hidden;
+
+        &::before {
+          border: var(--mdc-ripple-surface-test-edge-var);
+        }
+      }
+    }
+  }
+}
+
+@mixin mdc-ripple-surface($query: mdc-feature-all()) {
+  $feat-animation: mdc-feature-create-target($query, animation);
+  $feat-structure: mdc-feature-create-target($query, structure);
+
+  @include mdc-feature-targets($feat-structure) {
+    --mdc-ripple-fg-size: 0;
+    --mdc-ripple-left: 0;
+    --mdc-ripple-top: 0;
+    --mdc-ripple-fg-scale: 1;
+    --mdc-ripple-fg-translate-end: 0;
+    --mdc-ripple-fg-translate-start: 0;
+
+    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
+    will-change: transform, opacity;
```

Comment:
`will-change` should be moved to `animation` feature set? or is it used in both `structure` & `animation` feature sets?

Reply:
This one seems like it could go in either/or, but I've moved it to animation.

Reply:
Pat made a good point that `will-change` can affect appearance because it causes the content to render on a different layer, so I've moved this back to structure. We wouldn't want e.g. screenshot tests having differences from real use due to styles suppressed with the intent of skipping animations.